File: 351002/Lashchenka/src/main/java/by/bsuir/distcomp/service/MarkService.java
package by.bsuir.distcomp.service;

import by.bsuir.distcomp.dto.request.MarkRequestTo;
import by.bsuir.distcomp.dto.response.MarkResponseTo;
import by.bsuir.distcomp.entity.Mark;
import by.bsuir.distcomp.exception.DuplicateException;
import by.bsuir.distcomp.exception.ResourceNotFoundException;
import by.bsuir.distcomp.mapper.MarkMapper;
import by.bsuir.distcomp.repository.impl.InMemoryMarkRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class MarkService {

    private final InMemoryMarkRepository markRepository;
    private final MarkMapper markMapper;

    public MarkService(InMemoryMarkRepository markRepository, MarkMapper markMapper) {
        this.markRepository = markRepository;
        this.markMapper = markMapper;
    }

    public MarkResponseTo create(MarkRequestTo dto) {
        if (markRepository.existsByName(dto.getName())) {
            throw new DuplicateException("Mark with name '" + dto.getName() + "' already exists", 40305);
        }
        Mark entity = markMapper.toEntity(dto);
        Mark saved = markRepository.save(entity);
        return markMapper.toResponseDto(saved);
    }

    public MarkResponseTo getById(Long id) {
        Mark entity = markRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Mark with id " + id + " not found", 40409));
        return markMapper.toResponseDto(entity);
    }

    public List<MarkResponseTo> getAll() {
        return markRepository.findAll().stream()
                .map(markMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    public MarkResponseTo update(MarkRequestTo dto) {
        if (!markRepository.existsById(dto.getId())) {
            throw new ResourceNotFoundException("Mark with id " + dto.getId() + " not found", 40410);
        }
        if (markRepository.existsByNameAndIdNot(dto.getName(), dto.getId())) {
            throw new DuplicateException("Mark with name '" + dto.getName() + "' already exists", 40306);
        }
        Mark entity = markMapper.toEntity(dto);
        Mark updated = markRepository.update(entity);
        return markMapper.toResponseDto(updated);
    }

    public void deleteById(Long id) {
        if (!markRepository.existsById(id)) {
            throw new ResourceNotFoundException("Mark with id " + id + " not found", 40411);
        }
        markRepository.deleteById(id);
    }
}

File: 351003/Soika/entities-app/src/main/kotlin/com/example/entitiesapp/service/MarkService.kt
package com.example.entitiesapp.service

import com.example.entitiesapp.dto.MarkRequestTo
import com.example.entitiesapp.dto.MarkResponseTo
import com.example.entitiesapp.exception.NotFoundException
import com.example.entitiesapp.model.Mark
import com.example.entitiesapp.repository.MarkRepository
import org.springframework.stereotype.Service

@Service
class MarkService(
    private val repository: MarkRepository
) {

    private fun toEntity(dto: MarkRequestTo) = Mark(name = dto.name)
    private fun toResponse(entity: Mark) = MarkResponseTo(id = entity.id!!, name = entity.name)

    fun getAll(): List<MarkResponseTo> = repository.findAll().map { toResponse(it) }

    fun getById(id: Long): MarkResponseTo =
        toResponse(
            repository.findById(id)
                ?: throw NotFoundException("Mark with id=$id not found", 40402)
        )

    fun create(dto: MarkRequestTo): MarkResponseTo =
        toResponse(repository.save(toEntity(dto)))

    fun update(id: Long, dto: MarkRequestTo): MarkResponseTo {
        if (repository.findById(id) == null)
            throw NotFoundException("Mark with id=$id not found", 40402)
        return toResponse(repository.update(id, toEntity(dto)))
    }

    fun delete(id: Long) {
        if (repository.findById(id) == null)
            throw NotFoundException("Mark with id=$id not found", 40402)
        repository.delete(id)
    }
}